src/protocol.rs

```diff
@@ -23,35 +33,55 @@ pub enum CompileResponse {
     UnhandledCompile,
 }
 
+/// Server statistics.
 #[derive(Serialize, Deserialize, Debug)]
 pub struct CacheStats {
+    /// A `Vec` of individual statistics.
     pub stats: Vec<CacheStatistic>,
 }
 
+/// A single server statistic.
 #[derive(Serialize, Deserialize, Debug)]
 pub struct CacheStatistic {
+    /// Stat name.
     pub name: String,
+    /// Stat value.
     pub value: CacheStat,
 }
 
+/// A statistic value.
 #[derive(Serialize, Deserialize, Debug, PartialEq)]
 pub enum CacheStat {
+    /// A count of occurrences.
     Count(u64),
+    /// An opaque string, such as a name.
     String(String),
+    /// A size in bytes.
     Size(u64),
 }
 
+/// Information about a finished compile, either from cache or executed locally.
 #[derive(Serialize, Deserialize, Debug, Default)]
 pub struct CompileFinished {
+    /// The return code of the compile process, if available.
     pub retcode: Option<i32>,
+    /// The signal that terminated the compile process, if available.
     pub signal: Option<i32>,
+    /// The compiler's stdout.
     pub stdout: Vec<u8>,
+    /// The compiler's stderr.
     pub stderr: Vec<u8>,
 }
 
+/// The contents of a compile request from a client.
 #[derive(Serialize, Deserialize, Debug)]
 pub struct Compile {
+    /// The full path to the compiler executable.
     pub exe: String,
+    /// The current working directory in which to execute the compile.
     pub cwd: String,
+    /// The commandline arguments passed to the compiler.
     pub args: Vec<String>,
+    /// The environment variables present when the compiler was executed, as (var, val).
+    pub env_vars: Vec<(String, String)>,
```

Comment:
I pushed the serde bump to master and then fell down this rabbit hole. The only thing that I didn't have a great solution for was adding `OsString`s to the hash, so [I wrote a crappy equivalent of `OsStringExt::as_bytes` for Windows](https://github.com/luser/sccache/blob/27d719873d3e7ffa3de69f8d387f10e89dd18875/src/util.rs#L62) that just `mem::transmute`s. I suspect it will not ever be a problem in practice (and I added a unit test that should fail if it ever breaks) but it sure *feels* wrong.